feat(syntax): Add syntax for Cangjie language

WuJunkai2004 opened the pull request:
About issues #18014

This PR introduces a new, comprehensive syntax highlighting file for the Cangjie programming language.

**What is Cangjie?**

Cangjie is a new-generation programming language developed by Huawei. It is designed for all-scenario intelligence and features high performance and strong security. More information can be found at [here](https://cangjie-lang.cn/).

**Summary of Changes**

This PR includes all necessary components for adding a new language syntax:

1.  **New File**: `runtime/syntax/cangjie.vim`
2.  **Filetype Detection**: Added a rule to `runtime/filetype.vim` to recognize the `.cj` extension.
3.  **Documentation**: Updated `runtime/doc/syntax.txt` to include `cangjie` in the help document.
4.  **Menus**: Some menus.

**Features of the Syntax File**

The provided syntax script is robust and covers all major aspects of the Cangjie language, following modern Vim syntax best practices:

-   **Keywords & Types**: Full coverage of keywords and built-in types.
-   **Contextual Highlighting**: `func` and `struct` keywords trigger special highlighting for the following type/function name.
-   **Numbers**: Supports integers, floats, scientific notation, hex, binary, and octal, including `_` separators.
-   **Strings**: Supports single, double, triple-quoted strings, raw strings, and character runes.
-   **Interpolation]**: Correctly handles `${...}` interpolation.
-   **Operators**: A comprehensive and correctly ordered list of operators to prevent conflicts.
-   **Attributes**: Highlights annotations like `@override`.

The file has been tested with a sample covering these features. This contribution will provide a high-quality out-of-the-box experience for Cangjie developers using Vim.

Thank you for your consideration. Love U

WuJunkai2004 committed:
syntax: Add syntax for Cangjie language

This commit introduces a new syntax highlighting file for the Cangjie
programming language, includes 4 parts as required:
- The main syntax file: runtime/syntax/cangjie.vim
- The filetype detection rule in: runtime/filetype.vim
- The documentation update in: runtime/doc/syntax.txt
- Some menus

zeertzjq commented on runtime/doc/syntax.txt:
```diff
@@ -1116,6 +1116,17 @@ in the "after" directory in 'runtimepath'.  For Unix this would be
     syn sync fromstart
     set foldmethod=syntax
 
+CANGJIE					*cangjie.vim* *ft-cangjie-syntax*
```
There should be a blank line after this line to match the help for other syntax files.

WuJunkai2004 replied:
thanks, my fault

WuJunkai2004 committed:
fix: lossing a blank line to match document

WuJunkai2004 committed:
fix: recommit the tags file

try to access all of check

zeertzjq commented on runtime/doc/syntax.txt:
```diff
@@ -1116,6 +1116,18 @@ in the "after" directory in 'runtimepath'.  For Unix this would be
+control the syntax highlighting of Cangjie.  Example: >
+	:let g:cangjie_keyword_color = 0
+This will disable the highlighting of keywords in Cangjie.
+See more highlighting variables in |cangjie.vim|.
```
Linking `|cangjie.vim|` here isn't useful. It just links back to the same place. It's better to reference `$VIMRUNTIME/syntax/cangjie.vim` here.

WuJunkai2004 replied:
thank you for your review.

WuJunkai2004 committed:
feat: change the document about cangjie

remove the useless link, replace with details

dkearns commented on runtime/syntax/cangjie.vim:
```diff
@@ -0,0 +1,151 @@
+endif
+
+let s:save_cpo = &cpo
+set cpo&vim
```
You don't need to set and restore `'cpo'` as you're not using `\` line continuations, unless I'm not seeing them.

WuJunkai2004 replied:
You're right, I'm not using `\` line continuations this moment. I've included it just for robustness, in case future modifications to the file introduce them.

dkearns replied:
This doesn't bother me but we've generally excluded it if not needed. @chrisbra?

chrisbra replied:
That’s a valid concern. I usually point it out in case it was copied from existing runtime files without much thought. But if it’s intentional, I don’t see a strong reason to insist on removing it, and would defer to the maintainer’s preference.

dkearns commented on runtime/syntax/cangjie.vim:
```diff
@@ -0,0 +1,151 @@
+syn match   cjComment /\v\/\/.*/				contains=cjTodo
+syn region  cjComment start=/\/\*/ end=/\*\//	contains=cjTodo,@Spell
+syn cluster cjCommentCluster contains=cjComment,cjTodo
+if s:enabled('comment')
```
We normally add all `:highlight` commands at the end of the file.  I don't think it's a hard rule though, and I understand this organisation and tend to do it in my personal syntax files as well.

WuJunkai2004 replied:
o Yes, I will move them to the end of the file later. Thank u

dkearns commented on runtime/syntax/cangjie.vim:
```diff
@@ -0,0 +1,151 @@
+syn case match
+
+" 1. comments
+syn keyword cjTodo	TODO FIXME XXX NOTE BUG contained
```
Can you please change the syntax group prefixes to match the filetype name?  So `s/cj/cangjie/`.

WuJunkai2004 replied:
Thank you for pointing this out. I had initially used `cj` thinking it was a clear enough abbreviation, it is an oversight. I will update all prefixes in next commit.

dkearns reviewed (changes requested):
Thanks for the addition.

Could you also add a filetype detection test to `src/testdir/test_filetype.vim`?

WuJunkai2004 committed:
feat: modify the files to make them more standard

WuJunkai2004 committed:
fix: the suffix setting error

fix the wrong suffix setting. the
correct suffix is `cj`, not `cangjie`

WuJunkai2004 committed:
feat: add some highlightings

the New highlightings is added in Cangjie 1.0.0
- add the suffix of number
- add the b'' string
- add the range operator which I forgot

dkearns reviewed (approved):
Thanks.
